in-progress/8509-prover-coordination/design.md

```diff
@@ -0,0 +1,124 @@
+| Approvers            | @aminsammara @LHerskind @spalladino                         |
+| Target Approval Date | 2024-09-18                                                  |
+
+
+## Executive Summary
+
+The aztec node will provide an interface for proving marketplaces to submit quotes for proving an epoch.
+
+A proposer will be able to respond to a quote with a signed message, which a prover can submit to L1, along with a prover commitment bond. The prover can subsequently submit the proof to L1, releasing the bond.
+
+This coordination will take place "out of protocol", via an optional p2p topic, with the slight exception/leak that the rollup contract will permit the prover to submit bonds/proofs in a proposer's stead.
+
+## Introduction
+
+Presently, there is no coordination between proposers or provers. A coordination mechanism is necessary to ensure that:
+
+1. Proposers are able to obtain proofs for epochs, considering they will likely not have proving infrastructure.
+2. Provers are not obliged to "race" each other to submit proofs, as this would be a waste of resources.
+
+The protocol only "cares" about the proofs being submitted, but in practice, the node will need to coordinate the submission of proofs in a timely, efficient manner.
+
+## Interface
+
+Proving marketplaces will run full nodes, which will follow the pending chain. The node will expose a json-rpc endpoint for `submitEpochProofBid`, with the following signature:
+
+```typescript
+interface ProverCoordination {
+  submitEpochProofBid(epoch: number, basisPointFee: number): Promise<void>;
+}
+```
+
+This will be exposed via the cli as `aztec prover-coordination submit-epoch-proof-bid --epoch 123 --basis-point-fee 1000`.
+
+Under the hood, the node will submit the following message to `/aztec/proof-quotes/make/0.1.0` 
+
+
+```solidity
+struct Quote {
+  address prover;
+  uint256 epoch;
+  uint32 basisPointFee;
+  Signature signature;
+}
+```
+
+The `signature` will be produced using the L1 private key defined in the environment variable `PROVER_PUBLISHER_PRIVATE_KEY` to sign the message `keccak256(abi.encode(prover, epoch, basisPointFee))`.,
+
+Proposers will counter-sign the quote by submitting a message to `/aztec/proof-quotes/take/0.1.0` with the following same `Quote` struct, but with the `signature` produced using the L1 private key defined in the environment variable `VALIDATOR_PRIVATE_KEY`.
+
+The prover will be able to submit this Quote to `claimProofRight` on the rollup contract. The rollup contract will verify that:
+- The epoch is in the proof claim phase
+- There is not already a claim/proof for this epoch
+- The quote has been signed by the proposer for the current slot
+- The `prover` matches `msg.sender`
+- The epoch matches the previous epoch (i.e. the one to be proven)
+- The prover successfully transmits the bond (in TST) to the rollup contract
+
+If all conditions are met, the rollup contract stores the quote, and the address of the proposer.
```

Comment:
What happens if...?
- The proposer signs multiple quotes. Should the first one to get to L1 win?
- A malicious agent spams the p2p layer with cheap fake quotes. The sequencer signs one of those, but then the prover never uploads anything. How can a sequencer differentiate this attack from a legitimate first-time prover with no reputation yet? Should we require whitelisting?

Reply:
taking multiple quotes: Yes, I think this can/should be expected. I think it is a great feature. If Alice is proposer in an epoch, and she sees 10 bids, she can/should sign the 3 that advertise a price she is comfortable with; from her perspective then it doesn't matter which one posts the bond. It then de-risks Alice if one of those provers does not post the bond.

fake quotes: comes back, I think, to validation (checking L1 balance of the prover before signing), and possibly peer scoring: we can ask nodes participating on this layer to check L1 balance.